### PR TITLE
Query : NodeInfo

### PR DESCRIPTION
**Pull request by parv0888**
## Purpose
Adds support for query `NodeInfo`

## Changes
Adds support for query `NodeInfo`

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

## CLA acceptance
By submitting the contribution I accept the terms and conditions of the
Contributor License Agreement v1.0
- link: https://developers.concordium.com/CLAs/Contributor-License-Agreement-v1.0.pdf

- [x] I accept the above linked CLA.

### CODE REVIEW

File: concordium-sdk/src/main/java/com/concordium/sdk/Client.java [hunk 1]
```diff
@@ -245,6 +246,18 @@ public CryptographicParameters getCryptographicParameters(Hash blockHash) throws
         return cryptographicParameters;
     }
 
+
+    /**
+     * Gets the Node information.
+     * @return Parsed {@link NodeInfo}
+     * @throws Exception
+     */
+    public NodeInfo getNodeInfo() throws Exception {
```
Comment (parv0888) [inline concordium-sdk/src/main/java/com/concordium/sdk/Client.java:255]: @EmilLa1  - should we create a special exception ?

**Review by EmilLa1 — CHANGES_REQUESTED** [review]
Looks generally OK. 
A few comments and I think we can simplify the `NodeInfo` quite a bit. 

See for example https://github.com/Concordium/concordium-java-sdk/pull/81

EDIT: Also please add it to the README.md

File: concordium-sdk/src/main/java/com/concordium/sdk/Client.java (hunk 1 above)
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/Client.java:255]: thanks

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/PeerDetails.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+/**
+ * Peer Details in {@link NodeInfo}.
+ */
+@RequiredArgsConstructor
+public class PeerDetails {
+
+    /**
+     * Type of Peer. Bootstrapper Or Node.
+     */
+    private final PeerType type;
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/PeerDetails.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeInfo.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+
+/**
+ * Node Information.
+ */
+@Builder
+@Getter
+public class NodeInfo {
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeInfo.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/CommitteeDetails.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * Represents Baking Committee details for a Node.
+ */
+@Builder
+@Getter
+public class CommitteeDetails {
+
+    /**
+     * Baker ID of the Node.
+     */
+    private final UInt64 bakerId;
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/CommitteeDetails.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/PeerDetailsNodeActive.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * Represents {@link PeerDetailsNode} of type {@link ConsensusState.Active}.
+ */
+public class PeerDetailsNodeActive extends PeerDetailsNode {
+
+    /**
+     * Type of State of the Active Node.
+     */
+    private final NodeActiveStateType activeStateType;
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/PeerDetailsNodeActive.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/PeerDetailsNodeActive.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
+package com.concordium.sdk.responses.nodeInfo;
+
+import concordium.ConcordiumP2PRpc;
+import lombok.Builder;
+
+import java.util.Optional;
+
+/**
+ * Represents {@link PeerDetailsNode} of type {@link ConsensusState.Active}.
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/PeerDetailsNodeActive.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/PeerDetailsNode.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
+package com.concordium.sdk.responses.nodeInfo;
+
+import concordium.ConcordiumP2PRpc;
+import lombok.Getter;
+import lombok.val;
+
+/**
+ * Represents PeerDetail of type {@link PeerType.Node}.
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/PeerDetailsNode.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/Client.java [hunk 8]
```diff
@@ -2,6 +2,7 @@
 
 import com.concordium.sdk.exceptions.*;
 import com.concordium.sdk.responses.AccountIndex;
+import com.concordium.sdk.responses.nodeInfo.NodeInfo;
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/Client.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/PeerDetailsNodeActive.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+                        .committeeDetails(Optional.empty())
+                        .activeStateType(NodeActiveStateType.AddedButNotActiveInCommittee)
+                        .build();
+            case ADDED_BUT_WRONG_KEYS:
+                return PeerDetailsNodeActive.builder()
+                        .committeeDetails(Optional.empty())
+                        .activeStateType(NodeActiveStateType.AddedButWrongKeys)
+                        .build();
+            default:
+            case UNRECOGNIZED:
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/PeerDetailsNodeActive.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/PeerType.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package com.concordium.sdk.responses.nodeInfo;
+
+/**
+ * Represents type of Node.
+ */
+public enum PeerType {
+    /**
+     * Node of type Bootstrapper.
+     */
+    Bootstrapper,
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/PeerType.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
+package com.concordium.sdk.responses.nodeInfo;
+
+/**
+ * Represents Baker nodes status.
+ */
+public enum NodeActiveStateType {
+    /**
+     * Node is not part of Baking Committee.
+     */
+    NotInCommittee,
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+    /**
+     * Node is part of Baking Committee but not active.
+     */
+    AddedButNotActiveInCommittee,
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+    /**
+     * Node is part of the Baking Committee but has the wrong keys.
+     */
+    AddedButWrongKeys,
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+
+    /**
+     * Node is active in the Baking Committee.
+     */
+    ActiveInCommittee,
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+    /**
+     * Unrecognized State.
+     */
+    Unrecognized,
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java [hunk 16]
```diff
@@ -0,0 +1,31 @@
+package com.concordium.sdk.responses.nodeInfo;
+
+/**
+ * Represents Baker nodes status.
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+     */
+    AddedButNotActiveInCommittee,
+
+    /**
+     * Node is part of the Baking Committee but has the wrong keys.
+     */
+    AddedButWrongKeys,
+
+    /**
+     * Node is active in the Baking Committee.
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+     */
+    NotInCommittee,
+
+    /**
+     * Node is part of Baking Committee but not active.
+     */
+    AddedButNotActiveInCommittee,
+
+    /**
+     * Node is part of the Baking Committee but has the wrong keys.
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+ * Represents Baker nodes status.
+ */
+public enum NodeActiveStateType {
+    /**
+     * Node is not part of Baking Committee.
+     */
+    NotInCommittee,
+
+    /**
+     * Node is part of Baking Committee but not active.
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java [hunk 20]
```diff
@@ -0,0 +1,31 @@
+package com.concordium.sdk.responses.nodeInfo;
+
+/**
+ * Represents Baker nodes status.
+ */
+public enum NodeActiveStateType {
+    /**
+     * Node is not part of Baking Committee.
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/NodeActiveStateType.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/ConsensusState.java [hunk 21]
```diff
@@ -0,0 +1,21 @@
+package com.concordium.sdk.responses.nodeInfo;
+
+/**
+ * Consensus State of the Node.
+ */
+public enum ConsensusState {
+    /**
+     * Node's not running Consensus.
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/ConsensusState.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/ConsensusState.java [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+     */
+    NotRunning,
+
+    /**
+     * Node is running Consensus but is not Baking.
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/ConsensusState.java]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/ConsensusState.java [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+     */
+    Passive,
+
+    /**
+     * Node is a Baker.
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeInfo/ConsensusState.java]: done

**Comment by parv0888** [comment]
> Looks generally OK. A few comments and I think we can simplify the `NodeInfo` quite a bit.
> 
> See for example #81
> 
> EDIT: Also please add it to the README.md

Hi @EmilLa1  - have simplified the logic for parsing. Kindly review

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeinfo/NodeInfo.java [hunk 24]
```diff
@@ -30,37 +32,89 @@ public class NodeInfo {
...
+    /**
+     * Type of Peer Consensus Status in Baking Committee.
+     */
+    private final BakingStatus bakingStatus;
+
+    /**
+     * Committee details. This will only be populated if {@link NodeInfo#bakingStatus}
+     * is {@link BakingStatus#ACTIVE_IN_COMMITTEE}.
+     */
+    private final BakingCommitteeDetails committeeDetails;
```
Comment (parv0888) [inline concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeinfo/NodeInfo.java:53]: @EmilLa1  - should we keep this optional ?

Comment (EmilLa1) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeinfo/NodeInfo.java:53]: Yes I think that was a good idea.

**Review by EmilLa1 — CHANGES_REQUESTED** [review]
Looks good, minor comments. 👍 

Also please rebase so we can merge.

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeinfo/NodeInfo.java [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    private static PeerType parsePeerType(ConcordiumP2PRpc.NodeInfoResponse value) {
+        switch (value.getPeerType()) {
+            case "Bootstrapper" :
+                return PeerType.BOOTSTRAPPER;
+            case "Node" :
+                return PeerType.NODE;
+            default: throw new IllegalArgumentException(
+                    String.format("Invalid Peer Type : %s", value.getPeerType()));
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeinfo/NodeInfo.java:79]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeinfo/PeerType.java [hunk 26]
```diff
@@ -0,0 +1,16 @@
+package com.concordium.sdk.responses.nodeinfo;
+
+/**
+ * Represents type of Node.
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeinfo/PeerType.java:4]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeinfo/PeerType.java [hunk 27]
```diff
@@ -0,0 +1,16 @@
...
+ */
+public enum PeerType {
+    /**
+     * Node of type Bootstrapper.
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/responses/nodeinfo/PeerType.java:8]: done

**Review by EmilLa1 — APPROVED** [review]
Looks good :+1:

**Pull request merged**